src/Geography.cpp

```diff
@@ -0,0 +1,236 @@
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <iostream>
+#include <sstream>
+#include "Geography.hpp"
+#include "ByteBuffer.hpp"
+
+namespace voltdb {
+void Geography::Ring::reverse()
+{
+    PointIterator rev_start = m_points.begin();
+    PointIterator rev_end   = m_points.end();
+    rev_start++;
+    rev_end--;
+    std::reverse(rev_start, rev_end);
+}
+
+bool Geography::Ring::approximatelyEqual(const Ring &rhs, double epsilon) const {
+    if (numPoints() != rhs.numPoints()) {
+        return false;
+    }
+    for (int idx = 0; idx < numPoints(); idx += 1) {
+        if (!GeographyPoint::approximatelyEqual(getPoint(idx), rhs.getPoint(idx), epsilon)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+std::string Geography::Ring::toString() const {
+    std::stringstream stream;
+    stream << "(";
+    std::string sep("");
+    for (int idx = 0; idx < numPoints(); idx += 1) {
+        const GeographyPoint &pt = getPoint(idx);
+        stream << sep
+               << pt.getLongitude()
+               << " "
+               << pt.getLatitude();
+        sep = ", ";
+    }
+    stream << ")";
+    return stream.str();
+}
+
+void Geography::Ring::serializeTo(ByteBuffer &buffer, bool reverseit) const {
+    buffer.putInt8(0);
+    buffer.putInt32(m_points.size() - 1);
+    int start = (reverseit ? m_points.size() - 1 : 0);
+    int end   = (reverseit ? 0 : m_points.size() - 1);
+    int delta = (reverseit ? -1 : 1);
+    for (int idx = start; idx != end; idx += delta) {
+        double x, y, z;
+        m_points[idx].getXYZCoordinates(x, y, z);
+        buffer.putDouble(x);
+        buffer.putDouble(y);
+        buffer.putDouble(z);
+    }
+    char zeros[38];
+    memset(zeros, 0, sizeof(zeros));
+    buffer.put(zeros, sizeof(zeros));
+}
+
+bool Geography::approximatelyEqual(const Geography &rhs, double epsilon) const {
+    if (numRings() != rhs.numRings()) {
+        return false;
+    }
+    for (int idx = 0; idx < m_rings.size(); idx += 1) {
+        if (!getRing(idx).approximatelyEqual(rhs.getRing(idx), epsilon)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+int32_t Geography::getSerializedSize() const {
+    // Null Geography values are
+    // just 4 bytes long - just enough
+    // for the -1.
+    if (isNull()) {
+        return 4;
+    }
+    int32_t answer
```

Comment:
why not initialize answer with 4 and for loop will get skipped when null and return answer. Also answer can be renamed to size or something that indicates its size.